samcli/lib/utils/hash.py

```diff
@@ -0,0 +1,60 @@
+"""
+Hash calculation utilities for files and directories.
+"""
+import os
+import hashlib
+
+BLOCK_SIZE = 4096
+
+
+def file_checksum(file_name):
+    """
+
+    Parameters
+    ----------
+    file_name: file name of the file for which md5 checksum is required.
+
+    Returns
+    -------
+    md5 checksum of the given file.
+
+    """
+    with open(file_name, "rb") as file_handle:
+        md5 = hashlib.md5()
+
+        # Save current cursor position and reset cursor to start of file
+        curpos = file_handle.tell()
+        file_handle.seek(0)
+
+        buf = file_handle.read(BLOCK_SIZE)
+        while buf:
+            md5.update(buf)
+            buf = file_handle.read(BLOCK_SIZE)
+
+        # Restore file cursor's position
+        file_handle.seek(curpos)
+
+        return md5.hexdigest()
+
+
+def dir_checksum(directory, followlinks=True):
+    """
+
+    Parameters
+    ----------
+    directory : A directory with an absolute path
+    followlinks: Follow symbolic links through the given directory
+
+    Returns
+    -------
+    md5 checksum of the directory.
+
+    """
+    md5_dir = hashlib.md5()
+    # Walk through given directory and find all directories and files.
+    for dirpath, _, filenames in os.walk(directory, followlinks=followlinks):
```

Comment:
Do you have to catch circular links?

Reply:
Not sure how os.walk handles that, can look into it.

Reply:
Tried to create a series of files that had a circular symlink reference.

```
 ~/w/ap/py37/s/.aws-sam/build/HelloWorldFunction ▓▒░ ls -lrth a                                                                                 ░▒▓ ✔ | py3.7 Py | 12:46:18
lrwxr-xr-x  1 srirammv  ANT\Domain Users     1B Feb 12 12:46 a -> c
 ~/w/ap/py37/s/.aws-sam/build/HelloWorldFunction ▓▒░ ls -lrth b                                                                                 ░▒▓ ✔ | py3.7 Py | 12:46:23
lrwxr-xr-x  1 srirammv  ANT\Domain Users     1B Feb 12 12:46 b -> a
 ~/w/ap/py37/s/.aws-sam/build/HelloWorldFunction ▓▒░ ls -lrth c                                                                                 ░▒▓ ✔ | py3.7 Py | 12:46:25
lrwxr-xr-x  1 srirammv  ANT\Domain Users     1B Feb 12 12:46 c -> b
```

Get this error:

```
Error: Unable to upload artifact HelloWorldFunction referenced by CodeUri parameter of HelloWorldFunction resource.
[Errno 62] Too many levels of symbolic links: '/Users/srirammv/workspace/apps/py37_app/sam-app/.aws-sam/build/HelloWorldFunction/a'
```

So Looks like circular references are taken care of.